### PR TITLE
[SPARK-14483][WEBUI] Display user name for each job and query

### PR DESCRIPTION
**Pull request by sarutak**
## What changes were proposed in this pull request?
#12123 (SPARK-14245) introduced a feature displaying user name who runs jobs on AllJobsPage.

User names are pulled from the system property "user.name" but the property means who runs an application and each jobs in an application can be run by different users.  
One example is the case of using ThriftServer.  
ThriftServer is run by a user but queries are not always submitted by the user.                                                                                

This patch resolve this issue and displaying user name for each job and query on WebUI.  
Following screenshots are the example of querying through ThriftServer run by root.                                                                           
- AllJobsPage(Before)

alljobspage-before
- AllJobsPage(After)

alljobspage
- JobPage(Before) 

jobpage-before
- JobPage(After) 

jobpage
- AllExecutionsPage(Before)

allexecutionspage-before
- AllExecutionsPage(After)

allexecutionpage
- ExecutionPage(Before) 

executionpage-before
- ExecutionPage(After)

executionpage

Of course, we can see user name when we use HistoryServer.    
## How was this patch tested?

Run some queries through ThriftServer and checked UI including HistoryServer.

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by syepes** [comment]
@sarutak This is a great addition to the original pull.

**Comment by sarutak** [comment]
Thanks :)

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
CC: @andrewor14 @tgravescs

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
@tgravescs If you have time, could you check this? Thanks.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
CC: @rxin because you are the original author of `ExecutionPage.scala` and `AllExecutionsPage.scala`.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
ping @tgravescs

**Comment by tgravescs** [comment]
sorry I haven't had time to get to this.  I haven't had much experience using thriftserver or other job servers that would allow this so it might be better for someone more familiar with that code path to review.

**Comment by sarutak** [comment]
@tgravescs No problem. I just asked you because you reviewed #12123. Thanks.

**Comment by sarutak** [comment]
retest this please.

**Comment by ajbozarth** [comment]
Looking through some stale prs to review and try to push through. This still looks like a good add, @sarutak could you rebase?

**Comment by tdas** [comment]
@sarutak this looks like a useful PR, mind rebasing to master?
Also a high level comment. The 'user' is hardly the second thing that most users want to see. Description is most important. So its best to add `user` column to the right of the `description`.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala [hunk 1]
```diff
@@ -31,8 +31,6 @@ private[ui] class JobsTab(parent: SparkUI) extends SparkUITab(parent, "jobs") {
   def isFairScheduler: Boolean =
     jobProgresslistener.schedulingMode == Some(SchedulingMode.FAIR)
 
-  def getSparkUser: String = parent.getSparkUser
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala]: why is this being removed?

File: project/MimaExcludes.scala [hunk 2]
```diff
@@ -771,7 +771,14 @@ object MimaExcludes {
         ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.annotation.AlphaComponent"),
         ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.annotation.Experimental"),
         ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.annotation.DeveloperApi")
+      ) ++ Seq(
+        // [SPARK-14483][WEBUI] Display user name foreach job and query
+        ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.SparkListenerJobStart.copy"),
```
Comment (tdas) [inline project/MimaExcludes.scala]: I think we can add the user field to JobStart without breaking backward compatibility. Please refactor the change such that mima excludes are not needed as much as possible.

**Comment by tdas** [comment]
@sarutak this is a good PR have. could you please resolve the conflicts with master?

**Comment by sarutak** [comment]
@ajbozarth  @tdas Thanks for your interest. I'll rebase this to master and reflect the comments.

**Comment by ajbozarth** [comment]
I only skimmed the code, but it seems good. I agree with @tdas about moving the user column to the right of the description column though. Also I believe your screen shot for JobPage(After) is the same as the before one. 

My main concern with this is clutter. This is extremely useful for applications like the thrift store but for cluster where every app only has only user this will just clutter the ui with duplicate info. Perhaps adding some `ui.multitenant` config that can toggle display for clusters/servers that this actually applies too?

**Comment by sarutak** [comment]
@tdas @ajbozarth Sorry for the late response. I'm attaching new screen captures as follows.

* AllJobsPage
alljobspage

* JobPage
jobpage

* SQLPage
sqlpage

* SQLDetailsPage
sqldetails

* ThriftServerPage
thriftpage

* SessionPage
sessionpage

**Comment by sarutak** [comment]
@ajbozarth Your concern seems to be reasonable but I think it might be better way than you suggested. One possible way is toggle the appearance of User column by checkbox on demand. I think it's better way but not the best.
@tdas Do you have any better idea?

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
Not my day...
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by ajbozarth** [comment]
The screen shots look great, and you're probably right about a better solution. I'm not sure if a checkbox is much better though, especially since theres no precedent for it on these pages.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
Hi all, where are we on this?

**Comment by HyukjinKwon** [comment]
gentle ping @ajbozarth @tdas

**Comment by ajbozarth** [comment]
I'm still worried about clutter like I mentioned before, but the additional table column is what I'm most worried about especially on the AllJobsPage. @sarutak how important is the user column on the AllJobsPage? I'd be ok with this as is if we got rid of that, but if theres a good argument for it we should probably add a column visibility toggle like we recently did on the ExecutorsPage

**Comment by HyukjinKwon** [comment]
gentle ping @sarutak on ^

**Comment by ajbozarth** [comment]
FYI #18331 has made some updates that affect this

**Comment by gatorsmile** [comment]
Really appreciate your contribution! Sorry, based on the comment, we might need to close this PR, but please submit more PRs in the future. Thanks again!

**Comment by sarutak** [comment]
Sorry for my late reply. I'll close this PR for now and might open another PR in the near future. Thanks!

**Pull request closed without merging**